mitmproxy/addons/maplocal.py

```diff
@@ -0,0 +1,110 @@
+import mimetypes
+import re
+import typing
+import urllib
+from pathlib import Path
+
+from mitmproxy import exceptions
+from mitmproxy import ctx
+from mitmproxy import http
+from mitmproxy.addons.modifyheaders import parse_modify_spec, ModifySpec
+
+
+class MapLocal:
+    def __init__(self):
+        self.replacements: typing.List[ModifySpec] = []
+
+    def load(self, loader):
+        loader.add_option(
+            "map_local", typing.Sequence[str], [],
+            """
+            Replacement pattern of the form "[/flow-filter]/regex/file-or-directory", where
+            the separator can be any character. The @ allows to provide a file path that
+            is used to read the replacement string.
+            """
+        )
+
+    def configure(self, updated):
+        if "map_local" in updated:
+            self.replacements = []
+            for option in ctx.options.map_local:
+                try:
+                    spec = parse_modify_spec(option, True, True)
+                except ValueError as e:
+                    raise exceptions.OptionsError(f"Cannot parse map_local option {option}: {e}") from e
+
+                self.replacements.append(spec)
+
+    def construct_candidate_path(self, base_path, path_components, filename):
+        candidate_path = base_path.joinpath("/".join(path_components + [filename]))
+        return str(candidate_path)
+
+    def sanitize_candidate_path(self, candidate_path, base_path):
+        try:
+            candidate_path = candidate_path.resolve(strict=True)
+            if base_path == candidate_path or base_path in candidate_path.parents:
+                return candidate_path
+        except FileNotFoundError:
+            pass
+        return None
+
+    def file_candidates(self, url: str, spec: ModifySpec) -> typing.List[Path]:
+        replacement = spec.replacement
+        candidates = []
+
+        if replacement.is_file():
+            candidates.append(replacement)
+
+        elif replacement.is_dir():
```

Comment:
Good idea!

Reply:
Hm, I see your point in removing I/O, but doesn't this complicate testing due to different path separators being used on Unix vs. Windows? I.e., on Windows we might want to test `:jpg:c:\\repl`, while on Linux it would be `:jpg:/repl`.
Thus, to have real cross-platform tests, we need to have OS-level parametrization. Maybe I'm over-complicating stuff.
Will look at this more closely in the evening.

Reply:
We can probably use `./static` as our test path. That should behave the same way cross-platform.

Reply:
Sure. I didn't know if we aim for tests that explicitly test differences between Unix and Windows.